Rocket.Chat/Controllers/Chat/MessagesViewController.swift

```diff
@@ -52,27 +49,44 @@ final class MessagesViewController: RocketChatViewController {
 }
 
 extension MessagesViewController {
+
+    override func collectionView(_ collectionView: UICollectionView, willDisplay cell: UICollectionViewCell, forItemAt indexPath: IndexPath) {
+        if viewModel.numberOfSections - indexPath.section < 5 {
+            viewModel.fetchMessages(from: viewModel.oldestMessageDateBeingPresented)
+        }
+
+        super.collectionView(collectionView, willDisplay: cell, forItemAt: indexPath)
+    }
+
     func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
-        let section = data[indexPath.section]
-        let viewModel = section.viewModels()[indexPath.row]
-        if let height = heightCache[viewModel.differenceIdentifier] {
+        guard let section = viewModel.itemAt(indexPath) else {
+            return .zero
+        }
+
+        let sectionViewModel = section.viewModels()[indexPath.row]
+        if let height = viewSizingModel.height(for: sectionViewModel.differenceIdentifier) {
             return CGSize(width: UIScreen.main.bounds.width, height: height)
         } else {
             let sizingCell = BasicMessageCell.sizingCell
             sizingCell.prepareForReuse()
-            sizingCell.viewModel = viewModel
+            sizingCell.viewModel = sectionViewModel
             sizingCell.configure()
             sizingCell.setNeedsLayout()
             sizingCell.layoutIfNeeded()
+
             let size = sizingCell.contentView.systemLayoutSizeFitting(UIView.layoutFittingCompressedSize)
-            heightCache[viewModel.differenceIdentifier] = size.height
+            viewSizingModel.set(height: size.height, for: sectionViewModel.differenceIdentifier)
```

Comment:
To avoid referencing `UIScreen.main` here: 

```
return CGSize(width: UIScreen.main.bounds.width, height: height)
``` 

We could cache the `CGSize` instead of the `height`. We just need to use `NSValue(cgSize:_)` and adjust the `set(height:_, for:_)` method to support a CGSize.

Then to retrieve from the cache we just call `nsValue.cgSizeValue`. This is something I was going to do but since my caching mechanism was temporary I wind up not doing.